packages/plugins/typed-modal/src/Modal.js

```diff
@@ -0,0 +1,366 @@
+    this.focusableEls = this.htmlDiv_.querySelectorAll('a[href],' +
+        'area[href], input:not([disabled]), select:not([disabled]),' +
+        'textarea:not([disabled]), button:not([disabled]), [tabindex="0"]');
+    Blockly.WidgetDiv.show(this, this.workspace_.RTL,
+        () => this.widgetDispose_());
+    this.widgetCreate_();
+    if (this.focusableEls.length > 0) {
+      this.focusableEls[0].focus();
+    }
+  }
+
+  /**
+   * Hide the Blockly modal.
+   */
+  hide() {
+    Blockly.WidgetDiv.hide();
+  }
+
+  /**
+   * The function to be called when the user hits the 'x' button.
+   * @protected
+   */
+  onCancel_() {
+    this.hide();
+  }
+
+  /**
+   * Add the Blockly modal to the widget div and position it properly.
+   * @protected
+   */
+  widgetCreate_() {
+    const widgetDiv = Blockly.WidgetDiv.DIV;
+
+    widgetDiv.style.width = '100%';
+    widgetDiv.style.height = '100%';
+    widgetDiv.style.left = '0px';
+    widgetDiv.style.top = '0px';
+    widgetDiv.style.position = 'fixed';
+    if (this.closeOnClick) {
+      this.hideEvent = this.addEvent_(widgetDiv, 'click', this, this.hide);
+      this.modalEvent = this.addEvent_(this.htmlDiv_, 'click', this, (e) => {
+        e.stopPropagation();
+      });
+    }
+    Blockly.utils.dom.addClass(this.htmlDiv_, 'blockly-modal-open');
+    widgetDiv.appendChild(this.htmlDiv_);
+  }
+
+  /**
+   * Disposes of any events or dom-references belonging to the editor.
+   * @protected
+   */
+  widgetDispose_() {
+    const widgetDiv = Blockly.WidgetDiv.DIV;
+    widgetDiv.style.width = 'auto';
+    widgetDiv.style.height = 'auto';
+    if (this.closeOnClick) {
+      Blockly.unbindEvent_(this.hideEvent);
+      Blockly.unbindEvent_(this.modalEvent);
+    }
+    Blockly.utils.dom.removeClass(this.htmlDiv_, 'blockly-modal-open');
+  }
+
+  /**
+   * Handle when the user goes to the previous focusable element.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleBackwardTab_(e) {
+    if (document.activeElement === this.focusableEls[0]) {
+      e.preventDefault();
+      this.focusableEls[this.focusableEls.length - 1].focus();
+    }
+  }
+
+  /**
+   * Handle when the user goes to the next focusable element.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleForwardTab_(e) {
+    const focusedElements = this.focusableEls;
+    if (document.activeElement === focusedElements[focusedElements.length - 1]) {
+      e.preventDefault();
+      this.focusableEls[0].focus();
+    }
+  }
+
+  /**
+   * Handles keydown event for a Blockly modal. Handles forward tab, backward
+   * tab, and escape button.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleKeyDown_(e) {
+    if (e.keyCode === Blockly.utils.KeyCodes.TAB) {
+      if (this.focusableEls.length <= 1) {
+        e.preventDefault();
+        e.stopPropagation();
+        return;
+      }
+      if (e.shiftKey) {
+        this.handleBackwardTab_(e);
```

Comment:
Seems like we're doing more work than we need to here. You're taking a snapshot of all the focusable elements, and then handling tab navigation on each and every one of them, rather: 

Couldn't you just handle "wrap" tab navigation for the first and last element, to tab lock and ensure the user doesn't end up outside of the modal.

More on that here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role

Reply:
I think that is what I am doing. The handleForwardTab and handlBackwardTab only check if I need to wrap, if I don't then they don't do anything. 
Looking at it more I don't need to hold all the focusableEls so updated to have a firstFocusableEl and lastFocusableEl